### PR TITLE
docs: add security best practices

### PR DESCRIPTION
**Pull request by alxflw**
Closes https://github.com/onflow/flow/issues/1063

Adding security best practices
______

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [ ] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

**Review by joshuahannan — COMMENTED** [review]
Where did these come from and where will this go in the doc site. There is a lot of overlap between these and the [design patterns](https://github.com/onflow/cadence/blob/master/docs/design-patterns.mdx) and [anti patterns](https://github.com/onflow/cadence/blob/master/docs/anti-patterns.mdx) docs, so I don't know if it is worth adding a new doc here instead of just updating the other ones with anything that is relevant from this.

**Comment by alxflw** [comment]
I'm just executing on a request to add the doc linked to from the original issue. If you want to know more, we need context from @narangchirag, @turbolent and @franklywatson

**Comment by narangchirag** [comment]
I am okay adding another section with "Security best practices". Even if there's some overlap with other sections its better to have a dedicated page for security

**Comment by alxflw** [comment]
pinging reviewers to get this approved and merged @turbolent @SupunS @dsainati1 @joshuahannan

File: docs/security-best-practices.mdx [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+This is an opinionated list of best practices Cadence developers should follow to write more secure Cadence code.
+
+Some practices listed below might overlap with advice in the [Cadence Anti-Patterns](/cadence/design-patterns) section, which is a recommended read as well.
+
+## References
+
+[References](/cadence/language/references) are ephemeral values and cannot be stored. If persistence is required, store a capability and borrow it when needed.
+
+Authorized references (references with the auth keyword) allow downcasting, e.g. a restricted type to its unrestricted type and should only be used in some specific cases.
```
Comment (joshuahannan) [inline docs/security-best-practices.mdx:13]:
```suggestion
Authorized references (references with the `auth` keyword) allow downcasting, e.g. a restricted type to its unrestricted type and should only be used in some specific cases.
```

File: docs/security-best-practices.mdx [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+When exposing functionality, provide the least access necessary. Do not use authorized references, as they can be downcasted, potentially allowing a user to gain access to supposedly restricted functionality. For example, the fungible token standard provides an interface to get the balance of a vault, without exposing the withdrawal functionality.
+
+Be aware that the subtype or unrestricted type could expose functionality that was not intended to be exposed. Do not use authorized references when exposing functionality. For example, the fungible token standard provides an interface to get the balance of a vault, without exposing the withdrawal functionality.
+
+## Account Storage
+
+Don't trust the users’ [account storage](/cadence/language/accounts#account-storage). Users have full control over their data and may reorganize it as they see fit. Users may store values in any path, so paths may store values of “unexpected” types. These values may be instances of types in contracts that the user deployed.
```
Comment (joshuahannan) [inline docs/security-best-practices.mdx:21]:
```suggestion
Don't trust a users’ [account storage](/cadence/language/accounts#account-storage). Users have full control over their data and may reorganize it as they see fit. Users may store values in any path, so paths may store values of “unexpected” types. These values may be instances of types in contracts that the user deployed.
```

File: docs/security-best-practices.mdx [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+Always [borrow](/cadence/language/capability-based-access-control) with the specific type that is expected. Or, check if the value is an instance of the expected type.
+
+## Auth Accounts
+
+Access to an AuthAccount gives full access to the storage, keys, and contracts. Therefore, [avoid using AuthAccount](cadence/anti-patterns#problem) as a function parameter unless absolutely necessary.
```
Comment (joshuahannan) [inline docs/security-best-practices.mdx:27]:
```suggestion
Access to an `AuthAccount` gives full access to the account's storage, keys, and contracts. Therefore, [avoid using AuthAccount](cadence/anti-patterns#problem) as a function parameter unless absolutely necessary.
```

File: docs/security-best-practices.mdx [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+It is preferable to use capabilities over direct AuthAccount storage when exposing account data. Using capabilities allows the revocation of access by unlinking and limits the access to a single value with a certain set of functionality – access to an AuthAccount gives full access to the whole storage, as well as key and contract management.
```
Comment (joshuahannan) [inline docs/security-best-practices.mdx:29]:
```suggestion
It is preferable to use capabilities over direct `AuthAccount` storage when exposing account data. Using capabilities allows the revocation of access by unlinking and limits the access to a single value with a certain set of functionality – access to an `AuthAccount` gives full access to the whole storage, as well as key and contract management.
```

File: docs/security-best-practices.mdx [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+## Capabilities
+
+Don’t store anything under the [public capability storage](/cadence/language/capability-based-access-control) unless strictly required. Anyone can access your public capability using `getCapability`. If something needs to be stored under public make sure only read functionality is provided by restricting its type using either a resource interface or struct interface.
```
Comment (joshuahannan) [inline docs/security-best-practices.mdx:33]:
```suggestion
Don’t store anything under the [public capability storage](/cadence/language/capability-based-access-control) unless strictly required. Anyone can access your public capability using `AuthAccount.getCapability`. If something needs to be stored under `/public/`, make sure only read functionality is provided by restricting its type using either a resource interface or struct interface.
```

File: docs/security-best-practices.mdx [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+When linking a capability the link might be already present. In that case, Cadence will not panic with a runtime error but the link function will return nil.
```
Comment (joshuahannan) [inline docs/security-best-practices.mdx:35]:
```suggestion
When linking a capability, the link might already be present. In that case, Cadence will not panic with a runtime error but the link function will return `nil`.
```

File: docs/security-best-practices.mdx [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+It is a good practice to check if the link does already exist before creating it with `getLinkTarget`. This function will return nil if the link does not exist.
```
Comment (joshuahannan) [inline docs/security-best-practices.mdx:37]:
```suggestion
It is a good practice to check if the link already exists with `getLinkTarget` before creating it. This function will return `nil` if the link does not exist.
```

File: docs/security-best-practices.mdx [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+If it is necessary to handle the case where borrowing a capability might fail, the check function can be used to verify that the target exists and has a valid type.
```
Comment (joshuahannan) [inline docs/security-best-practices.mdx:39]:
```suggestion
If it is necessary to handle the case where borrowing a capability might fail, the `account.check` function can be used to verify that the target exists and has a valid type.
```

File: docs/security-best-practices.mdx [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+Ensure capabilities cannot be accessed by unauthorized parties. For example, capabilities should not be accessible through a public field, including public dictionaries or arrays. Exposing a capability in such a way allows anyone to borrow it and perform all actions that the capability allows.
+
+## Transactions
+
+Audits of Cadence code should also include [transactions](/cadence/language/transactions), as they may contain arbitrary code, just, like in contracts. In addition, they are given full access to the accounts of the transaction’s signers, i.e. the transaction is allowed to manipulate the signers’ account storage, contracts, and keys.
+
+Signing a transaction gives access to the AuthAccount, i.e. full access to the account’s storage, keys, and contracts.
```
Comment (joshuahannan) [inline docs/security-best-practices.mdx:47]:
```suggestion
Signing a transaction gives access to the `AuthAccount`, i.e. full access to the account’s storage, keys, and contracts.
```

File: docs/security-best-practices.mdx [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+## Types
+
+Use [restricted types and interfaces](/cadence/language/restricted-types). Always use the most specific type possible, following the principle of least privilege. Types should always be as restrictive as possible, especially for resource types.
+
+If given a less-specific type, cast to the more specific type that is expected. For example, when implementing the fungible token standard, a user may deposit any fungible token, so the implementation should cast to the expected concrete fungible token type.
+
+## Access Control
+
+Declaring a field as [pub/access(all)](/cadence/language/access-control) only protects from replacing the field’s value, but the value itself can still be mutated if it is mutable. Remember that containers, like dictionaries, and arrays, are mutable.
```
Comment (joshuahannan) [inline docs/security-best-practices.mdx:59]:
```suggestion
Declaring a field as [`pub/access(all)`](/cadence/language/access-control) only protects from replacing the field’s value, but the value itself can still be mutated if it is mutable. Remember that containers, like dictionaries, and arrays, are mutable.
```

File: docs/security-best-practices.mdx [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+
+Prefer non-public access to a mutable state. That state may also be nested. For example, a child may still be mutated even if its parent exposes it through a field with non-settable access.
+
+Do not use the pub/access(all) modifier on fields and functions unless necessary. Prefer priv/access(self), or access(contract) and access(account) when other types in the contract or account need to have access.
```
Comment (joshuahannan) [inline docs/security-best-practices.mdx:63]:
```suggestion
Do not use the `pub/access(all)` modifier on fields and functions unless necessary. Prefer `priv/access(self)`, or `access(contract)` and `access(account)` when other types in the contract or account need to have access.
```

**Review by joshuahannan — APPROVED** [review]

**Pull request merged**